File: 1253-sort-the-matrix-diagonally/1253-sort-the-matrix-diagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        Map<Integer, ArrayList<Integer>> map = new HashMap<>();
        int rows = mat.length;
        int cols = mat[0].length;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int key = i-j;
                if(!map.containsKey(key)){
                    map.put(key,new ArrayList<>());
                }
                map.get(key).add(mat[i][j]);
            }
        }

        
        for (ArrayList<Integer> list : map.values()) {
            Collections.sort(list);
        }

       
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int key = i - j;
                mat[i][j] = map.get(key).remove(0);
            }
        }

        return mat;
    }
}

File: 1253-sort-the-matrix-diagonally/1253-sort-the-matrix-diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        unordered_map<int,vector<int>> mp;
        int r=mat.size();
        int c=mat[0].size();
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                mp[i-j].push_back(mat[i][j]);
            }
        }

        for(auto&it:mp){
            sort(it.second.begin(),it.second.end());
        }
        for(int i=r-1;i>=0;i--){
            for(int j=c-1;j>=0;j--){
                mat[i][j]=mp[i-j].back();
                mp[i-j].pop_back();
            }
         
        }
        return mat;
    }
};